File: client/java-evosql-memory/src/main/java/io/evolutiondb/evosqlmemory/Connection.java
// EvoSQL agent-memory connection (Java).
//
// Threading: a Connection is NOT thread-safe — keep one per thread.
// The C SDK's last-error storage is thread-local, so an error on
// one thread doesn't leak into another's reading.
package io.evolutiondb.evosqlmemory;

import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public final class Connection implements AutoCloseable {

    private Pointer handle;

    public Connection(String host, int port, String user, String password) {
        this.handle = Native.INSTANCE.evo_connect(host, port, user, password);
        if (this.handle == null) {
            throw raiseLast("evo_connect failed", -2);
        }
    }

    public int exec(String sql) {
        ensureOpen();
        int rc = Native.INSTANCE.evo_exec(handle, sql);
        if (rc < 0) throw raiseLast("evo_exec failed", rc);
        return rc;
    }

    public List<List<String>> query(String sql) {
        return query(sql, 256, 32, 4096);
    }

    public List<List<String>> query(String sql,
                                       int maxRows, int maxCols,
                                       int colBufSize) {
        ensureOpen();
        byte[] buf = new byte[maxRows * maxCols * colBufSize];
        IntByReference colsOut = new IntByReference(0);
        int n = Native.INSTANCE.evo_query(handle, sql, buf,
                                              maxRows, maxCols, colBufSize,
                                              colsOut);
        if (n < 0) throw raiseLast("evo_query failed", n);
        int cols = colsOut.getValue();
        List<List<String>> rows = new ArrayList<>(n);
        for (int r = 0; r < n; r++) {
            List<String> row = new ArrayList<>(cols);
            for (int c = 0; c < cols; c++) {
                int off = (r * maxCols + c) * colBufSize;
                row.add(readNUL(buf, off, colBufSize));
            }
            rows.add(row);
        }
        return rows;
    }

    public void memoryPut(String store, String ns, String key,
                            String valueJson, float[] embedding) {
        ensureOpen();
        String embLit = (embedding != null && embedding.length > 0)
                          ? formatVector(embedding) : null;
        int rc = Native.INSTANCE.evo_memory_put(handle, store, ns, key,
                                                    valueJson, embLit);
        if (rc != 0) throw raiseLast("memory_put failed", rc);
    }

    public void memoryPut(String store, String ns, String key,
                            String valueJson) {
        memoryPut(store, ns, key, valueJson, null);
    }

    public String memoryGet(String store, String ns, String key) {
        return memoryGet(store, ns, key, 32 * 1024);
    }

    public String memoryGet(String store, String ns, String key, int bufSize) {
        ensureOpen();
        byte[] buf = new byte[bufSize];
        int rc = Native.INSTANCE.evo_memory_get(handle, store, ns, key,
                                                    buf, bufSize);
        if (rc == -7) {
            throw new NotFoundException("memory_get: " + ns + "/" + key);
        }
        if (rc != 0) throw raiseLast("memory_get failed", rc);
        return readNUL(buf, 0, bufSize);
    }

    public void memoryDelete(String store, String ns, String key) {
        ensureOpen();
        int rc = Native.INSTANCE.evo_memory_delete(handle, store, ns, key);
        if (rc != 0) throw raiseLast("memory_delete failed", rc);
    }

    public void checkpointPut(String store, String threadId, String ns,
                                  String checkpointId,
                                  String valuesJson, String metadataJson) {
        ensureOpen();
        int rc = Native.INSTANCE.evo_checkpoint_put(handle, store, threadId,
                                                        ns, checkpointId,
                                                        valuesJson,
                                                        metadataJson);
        if (rc != 0) throw raiseLast("checkpoint_put failed", rc);
    }

    public String checkpointGetLatest(String store, String threadId) {
        return checkpointGetLatest(store, threadId, 64 * 1024);
    }

    public String checkpointGetLatest(String store, String threadId,
                                          int bufSize) {
        ensureOpen();
        byte[] buf = new byte[bufSize];
        int rc = Native.INSTANCE.evo_checkpoint_get_latest(handle, store,
                                                                threadId,
                                                                buf, bufSize);
        if (rc == -7) {
            throw new NotFoundException("checkpoint thread=" + threadId);
        }
        if (rc != 0) throw raiseLast("checkpoint_get_latest failed", rc);
        return readNUL(buf, 0, bufSize);
    }

    public static String formatVector(float[] vec) {
        if (vec == null || vec.length == 0) return "";
        byte[] buf = new byte[vec.length * 16 + 32];
        int n = Native.INSTANCE.evo_vector_format(vec, vec.length, buf, buf.length);
        if (n > 0) return readNUL(buf, 0, buf.length);
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < vec.length; i++) {
            if (i > 0) sb.append(',');
            sb.append(String.format(java.util.Locale.US, "%.6f", vec[i]));
        }
        return sb.append(']').toString();
    }

    @Override
    public void close() {
        if (handle != null) {
            Native.INSTANCE.evo_close(handle);
            handle = null;
        }
    }

    // ------------------------------------------------------------------
    // helpers
    // ------------------------------------------------------------------
    private void ensureOpen() {
        if (handle == null) {
            throw new IllegalStateException("Connection is closed");
        }
    }

    private static String readNUL(byte[] buf, int offset, int max) {
        int end = offset;
        int hardEnd = Math.min(buf.length, offset + max);
        while (end < hardEnd && buf[end] != 0) end++;
        return new String(buf, offset, end - offset, StandardCharsets.UTF_8);
    }

    private static EvoException raiseLast(String fallback, int code) {
        String msg   = Native.INSTANCE.evo_strerror(null);
        String state = Native.INSTANCE.evo_sqlstate(null);
        return new EvoException(msg == null || msg.isEmpty() ? fallback : msg,
                                 code,
                                 state == null ? "" : state);
    }
}

File: client/swift-evosql-memory/Sources/EvoSQLMemory/Connection.swift
// EvoSQL agent-memory connection (Swift).
//
// Threading: a Connection is NOT thread-safe — keep one per thread.
// Connection is a class (reference type) with a deinit that calls
// evo_close so a forgotten close still releases the C handle.
import Foundation
import CEvoSQLMemory

public struct EvoError: Error, CustomStringConvertible {
    public let message:  String
    public let code:     Int32
    public let sqlstate: String

    public var description: String {
        sqlstate.isEmpty ? message : "sql \(sqlstate): \(message)"
    }
}

public struct EvoNotFound: Error, CustomStringConvertible {
    public let message: String
    public var description: String { message }
}

public final class Connection {
    private var handle: OpaquePointer?

    public init(host: String, port: Int32,
                  user: String, password: String) throws {
        let raw = host.withCString { h in
            user.withCString { u in
                password.withCString { p in
                    evo_connect(h, port, u, p)
                }
            }
        }
        guard let raw = raw else {
            throw Self.lastError(fallback: "evo_connect failed", code: -2)
        }
        self.handle = OpaquePointer(raw)
    }

    deinit { close() }

    public func close() {
        if let h = handle {
            evo_close(UnsafeMutablePointer(h))
            handle = nil
        }
    }

    @discardableResult
    public func exec(_ sql: String) throws -> Int32 {
        try ensureOpen()
        let rc = sql.withCString { s in
            evo_exec(UnsafeMutablePointer(handle!), s)
        }
        if rc < 0 { throw Self.lastError(fallback: "evo_exec failed", code: rc) }
        return rc
    }

    public func query(_ sql: String,
                       maxRows: Int32 = 256,
                       maxCols: Int32 = 32,
                       colBufSize: Int32 = 4096) throws -> [[String]] {
        try ensureOpen()
        let total = Int(maxRows) * Int(maxCols) * Int(colBufSize)
        var buf   = [Int8](repeating: 0, count: total)
        var cols: Int32 = 0

        let n = try buf.withUnsafeMutableBufferPointer { bp -> Int32 in
            sql.withCString { s in
                evo_query(UnsafeMutablePointer(handle!), s,
                            bp.baseAddress, maxRows, maxCols, colBufSize, &cols)
            }
        }
        if n < 0 { throw Self.lastError(fallback: "evo_query failed", code: n) }

        var rows: [[String]] = []
        rows.reserveCapacity(Int(n))
        for r in 0..<Int(n) {
            var row: [String] = []
            row.reserveCapacity(Int(cols))
            for c in 0..<Int(cols) {
                let off = (r * Int(maxCols) + c) * Int(colBufSize)
                let slot = Array(buf[off..<min(off + Int(colBufSize), buf.count)])
                row.append(Self.cString(slot))
            }
            rows.append(row)
        }
        return rows
    }

    public func memoryPut(store: String, namespace: String, key: String,
                            value: String,
                            embedding: [Float] = []) throws {
        try ensureOpen()
        let embLit = embedding.isEmpty ? nil : Self.formatVector(embedding)
        let rc = store.withCString { s in
            namespace.withCString { ns in
                key.withCString { k in
                    value.withCString { v in
                        if let embLit = embLit {
                            return embLit.withCString { e in
                                evo_memory_put(UnsafeMutablePointer(self.handle!),
                                                s, ns, k, v, e)
                            }
                        }
                        return evo_memory_put(UnsafeMutablePointer(self.handle!),
                                                s, ns, k, v, nil)
                    }
                }
            }
        }
        if rc != 0 { throw Self.lastError(fallback: "memory_put failed", code: rc) }
    }

    public func memoryGet(store: String, namespace: String, key: String,
                            bufSize: Int32 = 32 * 1024) throws -> String {
        try ensureOpen()
        var buf = [Int8](repeating: 0, count: Int(bufSize))
        let rc = buf.withUnsafeMutableBufferPointer { bp -> Int32 in
            store.withCString { s in
                namespace.withCString { ns in
                    key.withCString { k in
                        evo_memory_get(UnsafeMutablePointer(self.handle!),
                                          s, ns, k, bp.baseAddress, bufSize)
                    }
                }
            }
        }
        if rc == -7 {
            throw EvoNotFound(message: "memory_get: \(namespace)/\(key)")
        }
        if rc != 0 { throw Self.lastError(fallback: "memory_get failed", code: rc) }
        return Self.cString(buf)
    }

    public func memoryDelete(store: String, namespace: String,
                                key: String) throws {
        try ensureOpen()
        let rc = store.withCString { s in
            namespace.withCString { ns in
                key.withCString { k in
                    evo_memory_delete(UnsafeMutablePointer(self.handle!),
                                          s, ns, k)
                }
            }
        }
        if rc != 0 { throw Self.lastError(fallback: "memory_delete failed", code: rc) }
    }

    public func checkpointPut(store: String, threadId: String,
                                  namespace: String, checkpointId: String,
                                  values: String,
                                  metadata: String = "{}") throws {
        try ensureOpen()
        let rc = store.withCString { s in
            threadId.withCString { t in
                namespace.withCString { ns in
                    checkpointId.withCString { cid in
                        values.withCString { v in
                            metadata.withCString { m in
                                evo_checkpoint_put(UnsafeMutablePointer(self.handle!),
                                                       s, t, ns, cid, v, m)
                            }
                        }
                    }
                }
            }
        }
        if rc != 0 {
            throw Self.lastError(fallback: "checkpoint_put failed", code: rc)
        }
    }

    public func checkpointGetLatest(store: String, threadId: String,
                                          bufSize: Int32 = 64 * 1024)
        throws -> String
    {
        try ensureOpen()
        var buf = [Int8](repeating: 0, count: Int(bufSize))
        let rc = buf.withUnsafeMutableBufferPointer { bp -> Int32 in
            store.withCString { s in
                threadId.withCString { t in
                    evo_checkpoint_get_latest(UnsafeMutablePointer(self.handle!),
                                                   s, t, bp.baseAddress, bufSize)
                }
            }
        }
        if rc == -7 {
            throw EvoNotFound(message: "checkpoint thread=\(threadId)")
        }
        if rc != 0 {
            throw Self.lastError(fallback: "checkpoint_get_latest failed", code: rc)
        }
        return Self.cString(buf)
    }

    public static func formatVector(_ vec: [Float]) -> String {
        guard !vec.isEmpty else { return "" }
        var buf = [Int8](repeating: 0, count: vec.count * 16 + 32)
        let n = buf.withUnsafeMutableBufferPointer { bp -> Int32 in
            vec.withUnsafeBufferPointer { vp in
                evo_vector_format(vp.baseAddress, Int32(vec.count),
                                     bp.baseAddress, Int32(buf.count))
            }
        }
        if n > 0 { return cString(buf) }
        return "[" + vec.map { String(format: "%.6f", $0) }.joined(separator: ",") + "]"
    }

    // ------------------------------------------------------------------
    // helpers
    // ------------------------------------------------------------------
    private func ensureOpen() throws {
        if handle == nil {
            throw EvoError(message: "connection is closed",
                              code: -1, sqlstate: "")
        }
    }

    private static func cString(_ buf: [Int8]) -> String {
        var bytes: [UInt8] = []
        for b in buf {
            if b == 0 { break }
            bytes.append(UInt8(bitPattern: b))
        }
        return String(decoding: bytes, as: UTF8.self)
    }

    private static func lastError(fallback: String, code: Int32) -> EvoError {
        let msg: String = {
            if let p = evo_strerror(nil) { return String(cString: p) }
            return fallback
        }()
        let state: String = {
            if let p = evo_sqlstate(nil) { return String(cString: p) }
            return ""
        }()
        return EvoError(message: msg.isEmpty ? fallback : msg,
                          code: code, sqlstate: state)
    }
}
